File: 1_Month_Preparation_Kit/1MPK_W3/1MPK_W3_Reverse_A_Linked_List.erl
-module(solution).
-export([main/0]).
-import(os, [getenv/1]).

-record(singlyLinkedListNode, {data, next}).

create_singly_linked_list([]) ->
    null;
create_singly_linked_list([H | T]) ->
    #singlyLinkedListNode{data = H, next = create_singly_linked_list(T)}.

print_singly_linked_list(L) ->
    print_singly_linked_list(L, []).

print_singly_linked_list(null, Acc) ->
    Acc;
print_singly_linked_list(L, Acc) ->
    print_singly_linked_list(L#singlyLinkedListNode.next, Acc ++ [L#singlyLinkedListNode.data]).

%
% Complete the 'reverse' function below.
%
% The function is expected to return an INTEGER_SINGLY_LINKED_LIST.
% The function accepts INTEGER_SINGLY_LINKED_LIST llist as parameter.
%

%
% For your reference:
%
% singlyLinkedListNode = {
%     data = int,
%     next = singlyLinkedListNode
% }
%
%

reverse(Llist) ->
    % Write your code here
    reverse(Llist, null).
    
reverse(null, Acc) -> Acc;
reverse(Node, Acc) ->
    NewAcc = #singlyLinkedListNode{data = Node#singlyLinkedListNode.data, next = Acc},
    reverse(Node#singlyLinkedListNode.next, NewAcc).

read_multiple_lines_as_list_of_strings(N) ->
    read_multiple_lines_as_list_of_strings(N, []).

read_multiple_lines_as_list_of_strings(0, Acc) ->
    lists:reverse(Acc);
read_multiple_lines_as_list_of_strings(N, Acc) when N > 0 ->
    read_multiple_lines_as_list_of_strings(N - 1, [string:chomp(io:get_line("")) | Acc]).

main() ->
    {ok, Fptr} = file:open(getenv("OUTPUT_PATH"), [write]),

    {Tests, _} = string:to_integer(string:chomp(io:get_line(""))),

    lists:foreach(fun(TestsItr) ->
        {LlistTempCount, _} = string:to_integer(string:chomp(io:get_line(""))),

        LlistTempTemp = read_multiple_lines_as_list_of_strings(LlistTempCount),

        LlistTemp = lists:map(fun(X) -> {I, _} = string:to_integer(X), I end, LlistTempTemp),

        Llist = create_singly_linked_list(LlistTemp),

        Llist1 = reverse(Llist),

        Llist1Traversal = print_singly_linked_list(Llist1),

        io:fwrite(Fptr, "~s~n", [lists:join(" ", lists:map(fun(X) -> integer_to_list(X) end, Llist1Traversal))]) end,
    lists:seq(1, Tests)),

    file:close(Fptr),

    ok.

File: 1_Month_Preparation_Kit/1MPK_W3/1MPK_W3_Reverse_A_Linked_List.hs
{-# LANGUAGE FlexibleInstances, UndecidableInstances, DuplicateRecordFields #-}

module Main where

import Control.Monad
import Data.Array
import Data.Bits
import Data.List
import Data.List.Split
import Data.Set
import Debug.Trace
import System.Environment
import System.IO
import System.IO.Unsafe

data SinglyLinkedListNode = SinglyLinkedListNode {
    nodeData :: Int,
    next :: SinglyLinkedListNode
} | SinglyLinkedListNodeNil

createSinglyLinkedList :: [Int] -> SinglyLinkedListNode
createSinglyLinkedList [] = SinglyLinkedListNodeNil
createSinglyLinkedList (x:xs) = SinglyLinkedListNode {
    nodeData = x,
    next = createSinglyLinkedList xs
}

instance {-# OVERLAPPING #-} Show (SinglyLinkedListNode, String) where
    show (SinglyLinkedListNodeNil, _) = ""
    show (SinglyLinkedListNode x SinglyLinkedListNodeNil, _) = show x
    show (SinglyLinkedListNode x xs, sep) = show x ++ sep ++ show(xs, sep)

--
-- Complete the 'reverse' function below.
--
-- The function is expected to return an INTEGER_SINGLY_LINKED_LIST.
-- The function accepts INTEGER_SINGLY_LINKED_LIST llist as parameter.
--

--
-- For your reference:
--
-- SinglyLinkedListNode {
--     nodeData :: Int
--     next :: SinglyLinkedListNode
-- }
--
--


-- function is pure and requires no do
-- reverse llist = do
--    -- Write your code here


reverse :: SinglyLinkedListNode -> SinglyLinkedListNode
reverse llist = reverseHelper llist SinglyLinkedListNodeNil
    where
        reverseHelper SinglyLinkedListNodeNil acc = acc
        reverseHelper (SinglyLinkedListNode x next) acc = 
            reverseHelper next (SinglyLinkedListNode x acc)


readMultipleLinesAsStringArray :: Int -> IO [String]
readMultipleLinesAsStringArray 0 = return []
readMultipleLinesAsStringArray n = do
    line <- getLine
    rest <- readMultipleLinesAsStringArray(n - 1)
    return (line : rest)

main :: IO()
main = do
    stdout <- getEnv "OUTPUT_PATH"
    fptr <- openFile stdout WriteMode

    tests <- readLn :: IO Int

    forM_ [1..tests] $ \tests_itr -> do
        llistTempCount <- readLn :: IO Int

        llistTempTemp <- readMultipleLinesAsStringArray llistTempCount
        let llistTemp = Data.List.map (read :: String -> Int) llistTempTemp

        let llist = createSinglyLinkedList llistTemp

        let llist1 = reverse llist

        hPutStrLn fptr $ show(llist1, " ")

    hFlush fptr
    hClose fptr
